source/extensions/http/cache/file_system_http_cache/cache_file_header_proto_util.cc

```diff
@@ -0,0 +1,94 @@
+#include "source/extensions/http/cache/file_system_http_cache/cache_file_header_proto_util.h"
+
+#include "absl/container/flat_hash_set.h"
+#include "absl/strings/str_cat.h"
+
+namespace Envoy {
+namespace Extensions {
+namespace HttpFilters {
+namespace Cache {
+namespace FileSystemHttpCache {
+
+CacheFileHeader mergeProtoWithHeadersAndMetadata(const CacheFileHeader& entry_headers,
+                                                 const Http::ResponseHeaderMap& response_headers,
+                                                 const ResponseMetadata& response_metadata) {
+  Http::ResponseHeaderMapPtr merge_headers = headersFromHeaderProto(entry_headers);
+  applyHeaderUpdate(response_headers, *merge_headers);
+  return makeCacheFileHeaderProto(entry_headers.key(), *merge_headers, response_metadata);
+}
+
+CacheFileHeader makeCacheFileHeaderProto(const Key& key,
+                                         const Http::ResponseHeaderMap& response_headers,
+                                         const ResponseMetadata& metadata) {
+  CacheFileHeader file_header;
+  *file_header.mutable_key() = key;
+  TimestampUtil::systemClockToTimestamp(metadata.response_time_,
+                                        *file_header.mutable_metadata_response_time());
+  response_headers.iterate([&file_header](const Http::HeaderEntry& header) {
+    CacheFileHeader::Header* h = file_header.add_headers();
+    h->set_key(std::string{header.key().getStringView()});
+    h->set_value(std::string{header.value().getStringView()});
+    return Http::HeaderMap::Iterate::Continue;
+  });
+  return file_header;
+}
+
+CacheFileTrailer makeCacheFileTrailerProto(const Http::ResponseTrailerMap& response_trailers) {
+  CacheFileTrailer file_trailer;
+  response_trailers.iterate([&file_trailer](const Http::HeaderEntry& trailer) {
+    CacheFileTrailer::Trailer* t = file_trailer.add_trailers();
+    t->set_key(std::string{trailer.key().getStringView()});
+    t->set_value(std::string{trailer.value().getStringView()});
+    return Http::HeaderMap::Iterate::Continue;
+  });
+  return file_trailer;
+}
+
+size_t headerProtoSize(const CacheFileHeader& proto) { return proto.SerializeAsString().size(); }
+
+Buffer::OwnedImpl bufferFromProto(const CacheFileHeader& proto) {
+  return Buffer::OwnedImpl{proto.SerializeAsString()};
```

Comment:
It's the full set of http headers not the small block of sizes (that's why I named the other one FixedHeaderBlock rather than just Header, to help distinguish the two types that are right next to each other a lot.)

I think the least-bad way to serialize a proto without an intermediate string is SerializeToOStream, which I think would be a pretty hefty chunk of extra code to make an ostream capable of writing to a Buffer::OwnedImpl. There's also SerializeToArray but that would require pre-allocating the buffer to the correct size, and measuring a proto isn't that much faster than serializing. How would you feel about a TODO for this? :)

Reply:
sure that's fine, or even just a note to that this will contain the full serialized headers -- I think even full headers is relatively bounded due to default limits in Envoy, so this might be OK, especially relative to the speed of file i/o.